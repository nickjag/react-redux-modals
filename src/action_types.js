export const SHOW = 'react-redux-modals/SHOW';
export const HIDE = 'react-redux-modals/HIDE';
export const RESET = 'react-redux-modals/RESET';
