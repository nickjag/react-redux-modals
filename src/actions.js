import { SHOW, HIDE, RESET } from './action_types';

export const showModal = (modalType, modalProps) => ({
  type: SHOW,
  payload: { modalType, modalProps }
});

export const hideModal = (modalType) => ({
  type: HIDE,
  payload: modalType
});

export const resetModals = () => ({
  type: RESET
});