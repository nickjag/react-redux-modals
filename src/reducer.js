import { SHOW, HIDE, RESET } from './action_types';

const initialState = {
  activeModals: [],
  currentLevel: 1
};

export default (state = initialState, action) => {
  switch (action.type) {
    
    case SHOW:
      return {
        ...state,
        activeModals: [
          ...state.activeModals.filter(modal => modal.type !== action.payload.type),
          {
            ...action.payload,
            level: state.currentLevel + 2
          }
        ],
        currentLevel: state.currentLevel + 2
      };

    case HIDE:
      return {
        ...state,
        activeModals: state.activeModals.filter(modal => modal.type !== action.payload),
        currentLevel: (state.activeModals.filter(modal => modal.type !== action.payload).length > 0 
          ? state.currentLevel : 1)
      }

    case RESET:
      return initialState;

    default:
      return state;
  }
}