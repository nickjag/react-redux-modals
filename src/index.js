import reducer from './reducer';
import ModalRoot from './components/root';
import ModalFactory from './components/factory';
import Wrapper from './components/wrapper';
import * as actions from './actions';
import * as actionTypes from './action_types';

export default Wrapper;
export { reducer, actions, actionTypes, ModalRoot, ModalFactory };
