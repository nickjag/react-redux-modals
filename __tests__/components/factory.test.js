import React from 'react';
import { Provider } from 'react-redux';
import { ModalFactory } from '../../src';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';

const middlewares = [];
const mockStore = configureStore(middlewares);

describe('modal factory', () => {
  test('initially renders without children', () => {

    const initialState = { modals: {} };
    const store = mockStore(initialState);
    const wrapper = mount(<Provider store={store}><ModalFactory /></Provider>);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('renders with children after action', () => {

    const initialState = {
      modals: { 
        activeModals: [
          { modalType: 'MODAL_TEST', modalProps: null, level: 3 } 
        ],
        currentLevel: 3
      }
    };
    const store = mockStore(initialState);
    const modalTypes = { 'MODAL_TEST': (props) => <div>Modal Test!</div> };
    const wrapper = mount(<Provider store={store}><ModalFactory modalTypes={modalTypes} /></Provider>);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('fires prop on children', () => {

    const mockOnClick = jest.fn();
    const initialState = {
      modals: {
        activeModals: [
          { modalType: 'MODAL_TEST', modalProps: {
            onSubmit: mockOnClick
          }, level: 3 } 
        ],
        currentLevel: 3
      }
    };
    const store = mockStore(initialState);
    const modalTypes = {'MODAL_TEST': (props) => <div><button onClick={props.onSubmit} /></div>};
    const wrapper = mount(<Provider store={store}><ModalFactory modalTypes={modalTypes} /></Provider>);

    wrapper.find('button').simulate('click');

    expect(mockOnClick.mock.calls.length).toEqual(1);
  });

});
