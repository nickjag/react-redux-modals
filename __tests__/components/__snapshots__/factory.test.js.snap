// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modal factory initially renders without children 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ModalFactory)>
    <ModalFactory
      dispatch={[Function]}
    />
  </Connect(ModalFactory)>
</Provider>
`;

exports[`modal factory renders with children after action 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ModalFactory)
    modalTypes={
      Object {
        "MODAL_TEST": [Function],
      }
    }
  >
    <ModalFactory
      activeModals={
        Array [
          Object {
            "level": 3,
            "modalProps": null,
            "modalType": "MODAL_TEST",
          },
        ]
      }
      currentLevel={3}
      dispatch={[Function]}
      modalTypes={
        Object {
          "MODAL_TEST": [Function],
        }
      }
    >
      <div>
        <MODAL_TEST
          key="MODAL_TEST"
          modalConfig={
            Object {
              "level": 3,
            }
          }
        >
          <div>
            Modal Test!
          </div>
        </MODAL_TEST>
      </div>
    </ModalFactory>
  </Connect(ModalFactory)>
</Provider>
`;
